packages/devextreme/js/__internal/core/state_manager/state_manager.ts

```diff
@@ -0,0 +1,165 @@
+  ) {
+    this.componentState = {};
+
+    this.valueContainerManagers = config.valueContainerManagers;
+    this.devToolsConnector = config.devToolsConnector;
+    this.logger = config.logger;
+    this.stateSourceSign = config.stateSourceSign;
+
+    this.init();
+  }
+
+  private init(): void {
+    this.devToolsConnector.onExternalAction((action) => {
+      this.logger.warn(`Handler for the '${action}' action is not implemented`);
+    });
+
+    this.devToolsConnector.connect();
+
+    this.logger.info('StateManager initialized');
+  }
+
+  trackStateOf(sourceData: StateManagementTypes.StateSource, sourceDataId?: string): void {
+    const preparedSourceDataId = sourceDataId ?? sourceData?.constructor?.name;
+
+    if (!sourceData) {
+      this.logger.error('State source cannot be null or undefined');
+      return;
+    }
+
+    if (this.componentState[preparedSourceDataId]) {
+      this.logger.debug(`State source with ID '${preparedSourceDataId}' is already tracked. Overwriting.`);
+    }
+
+    Object.entries(sourceData).forEach(([propertyName, propertyValue]) => {
+      if (this.hasValueContainerManagerFor(propertyValue)) {
+        if (!this.componentState[preparedSourceDataId]) {
+          this.componentState[preparedSourceDataId] = {};
+        }
+
+        const isObject = propertyValue !== undefined && propertyValue !== null
+          && typeof propertyValue === 'object';
+        this.componentState[preparedSourceDataId][propertyName] = isObject
+          ? new WeakRef(propertyValue) : propertyValue;
+
+        this.trackStateSourceChanges(
+          preparedSourceDataId,
+          propertyName,
+          propertyValue,
+        );
+      } else {
+        this.logger.debug(`No value container manager found for the '${propertyName}' property of the '${preparedSourceDataId}' state source`);
+      }
+    });
+  }
+
+  private trackStateSourceChanges(
+    stateId: string,
+    propertyName: string,
+    propertyValue: StateManagementTypes.MaybeValueContainer,
+  ): void {
+    const valueContainerManager = this.findValueContainerManagerFor(propertyValue);
+
+    if (!valueContainerManager) {
+      this.logger.debug(`No value container manager found for the '${propertyName}' property of the '${stateId}' state`);
+      return;
+    }
+
+    const fullPathToProperty = joinStatePath(stateId, propertyName);
+
+    try {
+      valueContainerManager.trackChanges(
+        (valueContainerChange: StateManagementTypes.ValueContainerChange) => {
+          const componentStateChange = {
+            ...valueContainerChange,
+            payload: { ...valueContainerChange.payload, path: fullPathToProperty },
+          };
+
+          const currentComponentState = this.getComponentState();
+
+          this.devToolsConnector
+            .sendAction(
+              valueContainerChange.actionType,
+              componentStateChange.payload,
+              currentComponentState,
```

Comment:
How is it different from `sourceData` that comes to the `trackStateOf` method?

I see how `sourceData` gets wrapped, becomes `this.componentState` and gets unwrapped, but I can't grasp how `this.componentState` can be modified.